### PR TITLE
Reconcile Milestones with Release Branch before release

### LINKED ISSUE #44037: Parse Release Branch Commit History for Release notes + milestones
### General Idea
- parse every commit in the release branch since the last release tag
- for every commit in the release branch
  - find the backport PR
  - trace the backport PR to the PR(s) to master
  - find any/all linked issues from the PR to master
- add milestones to all issues with a link to the release branch for the current next open milestone
- flag all closed issues in the milestone that can't be traced to the release branch

### Problems
- an issue [cannot have multiple milestones](https://github.com/orgs/community/discussions/4566), so if a bugfix is backported to multiple release branches, we can't rely on milestones to generate accurate release notes
- how to handle reverts?
- notify of open issues in milestone?

### PR DESCRIPTION
**Pull request by iethree**
> [!important]
> automation token needs manage project scope added

Closes https://github.com/metabase/metabase/issues/44037

### Description

Screen Shot 2024-07-10 at 8 17 23 AM

- checks every commit in the release to make sure it corresponds to an issue in the milestone or conforms to an exception:
  - issues with an earlier milestone already set (e.g. also backported to another release branch)
  - issues with excluded labels for CI, Tests, and Documentation
- any problems are added to a [project](https://github.com/orgs/metabase/projects/93/views/1) and the link to the project is posted in the releases channel 
- posts a link in the slack release thread, and uploads a log of all its decisions

### Todo:
- [x] parse tag history for last release tag in branch
- [x] add github action
- [x] trigger github action from build job
- [x] add slack message

### CODE REVIEW

File: .github/workflows/release-milestone-check.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+jobs:
+  check-milestone:
+    runs-on: ubuntu-22.04
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          sparse-checkout: release
+      - name: Prepare build scripts
+        run: cd ${{ github.workspace }}/release && yarn && yarn build
```
Comment (uladzimirdev) [inline .github/workflows/release-milestone-check.yml:24]: `yarn --cwd release build`

File: release/src/milestones.ts [hunk 2]
```diff
@@ -196,3 +265,313 @@ export async function setMilestoneForCommits({
...
+      }
+      issueNumbers.push(...(await getOriginalIssues({
+        github,
+        owner,
+        repo,
+        issueNumber: prNumber,
+      })));
+    }
+
+    const uniqueIssues = _.uniq(issueNumbers.filter(isNotNull));
```
Comment (uladzimirdev) [inline release/src/milestones.ts:339]: I don't mind, but sometimes it's new Set, sometimes underscore

File: release/src/milestones.ts [hunk 3]
```diff
@@ -196,3 +265,313 @@ export async function setMilestoneForCommits({
...
+  issuesInOlderMilestones: Set<number>,
+  milestoneIssueSet: Set<number>,
+  openMilestoneIssues: Issue[],
+  commitIssueMap: Record<string, number[]>,
+}) {
+  let log = `# ${version} Milestone Audit Log\n\n`;
+
+  log += `Parsing from ${lastTag} to ${commitHash} \n`;
+
+  log += '## Summary:\n';
```
Comment (uladzimirdev) [inline release/src/milestones.ts:471]: nit: you can drop `log +=` on new lines to improve readability

**Review by uladzimirdev — APPROVED** [review]
LGTM from code perspective

File: release/src/milestones.ts (hunk 2 above)
Comment (iethree) [reply release/src/milestones.ts:338]: sometimes I want a set, sometimes I want an array 🤷

**Pull request merged**